% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Genomic Inversions in scoli Alter Gene Expression: Code and Analysis},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\usepackage{xspace, xcolor, hyperref}

\title{Genomic Inversions in scoli Alter Gene Expression: Code and Analysis}
\author{}
\date{\vspace{-2.5em}}

\begin{document}
\maketitle

\newcommand{\ecol}{\textit{E.\,coli}\xspace}
\newcommand{\escoli}{\textit{Escherichia coli}\xspace}

This Notebook gives an overview of key code, commands, data, and
analysis used for the paper titles ``Genomic Inversions in
\textit{Escherichia coli}\xspaceAlter Gene Expression: Code and
Analysis'' by Daniella F Lato, Qing Zeng, and G Brian Golding
(\textcolour{red}{insert url to paper here}).

\#Sequence Data Information for the RNA-seq data used in this analysis
can be found in the
\href{run:./Supplementary_Material_Lato_Zeng_Golding_GenomicInversionsInEcoli.pdf}{Supplementary Material.}

\#Gene Expression Information for the RNA-seq data used in this analysis
can be found in the
\href{run:./Supplementary_Material_Lato_Zeng_Golding_GenomicInversionsInEcoli.pdf}{Supplementary Material.}
Pre-processing to combine the raw expression files was done with custom
\texttt{shell} scripts.

\#\#Normalization: Gene expression was normalized using the following
\texttt{R} code.

\end{document}
